From a12f23f0ba6829a076087cc97c9d2c24396562d7 Mon Sep 17 00:00:00 2001
From: Ross Wille <wille@freescale.com>
Date: Mon, 24 Nov 2008 15:30:22 -0700
Subject: [PATCH] ENGR00099613 MX25: Add CSPI3 and iomux for CSPI2 and 3

Add CSPI3 support for MX25 3stack platform.
Setup the pins for CSPI2 and CSPI3.
Remove pin ownership code for CSPI DMA since
it is not currently supported.

Signed-off-by: Ross Wille <wille@freescale.com>
---
 arch/arm/mach-mx25/devices.c          |   45 +++++++++++++++++++++++++++++---
 arch/arm/mach-mx25/mx25_3stack_gpio.c |    4 ---
 drivers/spi/Kconfig                   |    2 +-
 3 files changed, 41 insertions(+), 10 deletions(-)

diff --git a/arch/arm/mach-mx25/devices.c b/arch/arm/mach-mx25/devices.c
index 29dbc94..edffc49 100644
--- a/arch/arm/mach-mx25/devices.c
+++ b/arch/arm/mach-mx25/devices.c
@@ -185,15 +185,46 @@ static struct platform_device mxcspi2_device = {
 };
 #endif				/* CONFIG_SPI_MXC_SELECT2 */
 
+#ifdef CONFIG_SPI_MXC_SELECT3
+/*!
+ * Resource definition for the CSPI3
+ */
+static struct resource mxcspi3_resources[] = {
+	[0] = {
+	       .start = CSPI3_BASE_ADDR,
+	       .end = CSPI3_BASE_ADDR + SZ_4K - 1,
+	       .flags = IORESOURCE_MEM,
+	       },
+	[1] = {
+	       .start = MXC_INT_CSPI3,
+	       .end = MXC_INT_CSPI3,
+	       .flags = IORESOURCE_IRQ,
+	       },
+};
+
+/*! Platform Data for MXC CSPI3 */
+static struct mxc_spi_master mxcspi3_data = {
+	.maxchipselect = 4,
+	.spi_version = 7,
+};
+
+/*! Device Definition for MXC CSPI3 */
+static struct platform_device mxcspi3_device = {
+	.name = "mxc_spi",
+	.id = 2,
+	.dev = {
+		.release = mxc_nop_release,
+		.platform_data = &mxcspi3_data,
+		},
+	.num_resources = ARRAY_SIZE(mxcspi3_resources),
+	.resource = mxcspi3_resources,
+};
+#endif				/* CONFIG_SPI_MXC_SELECT3 */
+
 static inline void mxc_init_spi(void)
 {
-#ifdef CONFIG_SPI_MXC_DMA
-	spba_take_ownership(SPBA_CSPI2, SPBA_MASTER_A | SPBA_MASTER_C);
-	spba_take_ownership(SPBA_CSPI3, SPBA_MASTER_A | SPBA_MASTER_C);
-#else
 	spba_take_ownership(SPBA_CSPI2, SPBA_MASTER_A);
 	spba_take_ownership(SPBA_CSPI3, SPBA_MASTER_A);
-#endif
 
 #ifdef CONFIG_SPI_MXC_SELECT1
 	if (platform_device_register(&mxcspi1_device) < 0)
@@ -203,6 +234,10 @@ static inline void mxc_init_spi(void)
 	if (platform_device_register(&mxcspi2_device) < 0)
 		printk(KERN_ERR "Error: Registering the SPI Controller_2\n");
 #endif				/* CONFIG_SPI_MXC_SELECT2 */
+#ifdef CONFIG_SPI_MXC_SELECT3
+	if (platform_device_register(&mxcspi3_device) < 0)
+		printk(KERN_ERR "Error: Registering the SPI Controller_3\n");
+#endif				/* CONFIG_SPI_MXC_SELECT3 */
 }
 #else
 static inline void mxc_init_spi(void)
diff --git a/arch/arm/mach-mx25/mx25_3stack_gpio.c b/arch/arm/mach-mx25/mx25_3stack_gpio.c
index 85e50e8..9690fc9 100644
--- a/arch/arm/mach-mx25/mx25_3stack_gpio.c
+++ b/arch/arm/mach-mx25/mx25_3stack_gpio.c
@@ -508,7 +508,6 @@ void gpio_spi_active(int cspi_mod)
 		break;
 	case 1:
 		/* SPI2 */
-#if 0
 		mxc_request_iomux(MX25_PIN_LD12, MUX_CONFIG_ALT2); /*MOSI*/
 		mxc_request_iomux(MX25_PIN_LD13, MUX_CONFIG_ALT2); /*MISO*/
 		mxc_request_iomux(MX25_PIN_LD14, MUX_CONFIG_ALT2); /*SCLK*/
@@ -537,11 +536,9 @@ void gpio_spi_active(int cspi_mod)
 				    INPUT_CTL_PATH0);
 		mxc_iomux_set_input(MUX_IN_CSPI2_IPP_IND_SS1_B,
 				    INPUT_CTL_PATH0);
-#endif
 		break;
 	case 2:
 		/* SPI3 */
-#if 0
 		mxc_request_iomux(MX25_PIN_EB0, MUX_CONFIG_ALT6); /*SS0*/
 		mxc_request_iomux(MX25_PIN_EB1, MUX_CONFIG_ALT6); /*SS1*/
 		mxc_request_iomux(MX25_PIN_CS4, MUX_CONFIG_ALT6); /*MOSI*/
@@ -572,7 +569,6 @@ void gpio_spi_active(int cspi_mod)
 				    INPUT_CTL_PATH1);
 		mxc_iomux_set_input(MUX_IN_CSPI3_IPP_IND_SS3_B,
 				    INPUT_CTL_PATH0);
-#endif
 		break;
 	default:
 		break;
diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index 694e336..fedb8c1 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -222,7 +222,7 @@ config SPI_MXC_SELECT2
 
 config SPI_MXC_SELECT3
 	bool "CSPI3"
-	depends on SPI_MXC && (ARCH_MX3 || ARCH_MX27 || ARCH_MX37 || ARCH_MX51)
+	depends on SPI_MXC && (ARCH_MX3 || ARCH_MX27 || ARCH_MX25 || ARCH_MX37 || ARCH_MX51)
 	default n
 
 #
-- 
1.5.4.4

